### PR TITLE
[SYCL-MLIR] Improve warning and debug messages

### PR DESCRIPTION
**Pull request by victor-eds**
Use mlirclang::warning or llvm::dbgs when suitable.

Signed-off-by: Victor Perez <victor.perez@codeplay.com>

### CODE REVIEW

File: polygeist/tools/cgeist/Lib/CGExpr.cc [hunk 1]
```diff
@@ -2561,7 +2560,7 @@ BinOpInfo MLIRScanner::EmitBinOps(BinaryOperator *E, QualType PromotionType) {
 static void informNoOverflowCheck(LangOptions::SignedOverflowBehaviorTy SOB,
                                   llvm::StringRef OpName) {
   if (SOB != clang::LangOptions::SOB_Defined)
-    llvm::errs() << "Not emitting overflow-checked " << OpName << "\n";
+    mlirclang::warning() << "Not emitting overflow-checked " << OpName << "\n";
```
Comment (victor-eds) [inline polygeist/tools/cgeist/Lib/CGExpr.cc:2563]: As reported [here](https://github.com/intel/llvm/pull/7421#discussion_r1034062417) by @whitneywhtsang

File: polygeist/tools/cgeist/Lib/CGExpr.cc [hunk 2]
```diff
@@ -1570,13 +1572,11 @@ ValueCategory MLIRScanner::VisitCastExpr(CastExpr *E) {
       return ValueCategory(Nval, /*isReference*/ false);
     }
 
-#ifdef DEBUG
```
Comment (whitneywhtsang) [inline polygeist/tools/cgeist/Lib/CGExpr.cc:1573]: What's the motivation to remove these `#ifdef DEBUG`? Should we guard them by `LLVM_DEBUG`?

Comment (victor-eds) [reply polygeist/tools/cgeist/Lib/CGExpr.cc:1573]: In other points of the programs, constructs like these (printing involved values that will miss a following `assert`) are not guarded, so I just followed that approach. IMO, we could just remove all of these completely, i.e., no printing before asserts.

Comment (whitneywhtsang) [reply polygeist/tools/cgeist/Lib/CGExpr.cc:1573]: oic, didn't notice they `assert` after print, I am fine with those.
How about the ones at line 1721 and 1781, should we put them in `LLVM_DEBUG`?

Comment (whitneywhtsang) [reply polygeist/tools/cgeist/Lib/CGExpr.cc:1573]: I am fine removing the guards for the prints that assert after, as we shouldn't see them when compile successfully. My concern is on the prints that show up even if we compile successfully. e.g., the one at line 1721:
```
#ifdef DEBUG
    if (!Scalar.getType().isa<mlir::IntegerType>()) {
      E->dump();
      llvm::errs() << " Scalar: " << Scalar << "\n";
    }
#endif

    auto PrevTy = Scalar.getType().cast<mlir::IntegerType>();
    bool SignedType = true;
    if (const auto *Bit =
            dyn_cast<clang::BuiltinType>(&*E->getSubExpr()->getType())) {
...
```
Am I missing something?

Comment (victor-eds) [reply polygeist/tools/cgeist/Lib/CGExpr.cc:1573]: Sorry, I was not clear. We're checking type correctness in the `if` and then `cast`ing. `cast` will throw an error if the type is incorrect.

**Review by whitneywhtsang — APPROVED** [review]

File: polygeist/tools/cgeist/Lib/CGExpr.cc (hunk 2 above)
Comment (whitneywhtsang) [reply polygeist/tools/cgeist/Lib/CGExpr.cc:1573]: True, thanks for clarifying.

File: polygeist/tools/cgeist/Lib/CGExpr.cc [hunk 3]
```diff
@@ -1551,10 +1553,8 @@ ValueCategory MLIRScanner::VisitCastExpr(CastExpr *E) {
             }
         }
     auto SE = Visit(E->getSubExpr());
-#ifdef DEBUG
     if (!SE.val)
```
Comment (etiotto) [inline polygeist/tools/cgeist/Lib/CGExpr.cc:1556]: Naked trace. We need to place this one in a LLVM_DEBUG.

Comment (etiotto) [reply polygeist/tools/cgeist/Lib/CGExpr.cc:1556]: Even if the code asserts later on one can always compile with traces enabled to get more information, so IMO it still make sense to use LLVM_DEBUG. Same comment for other similar changes. Also wrapping debug code in LLVM_DEBUG makes it easily identifiable to the human eye.

**Review by etiotto — COMMENTED** [review]
Suggest using LLVM_DEBUG macro to wrap debug code (even is an assertion will be emitted).

File: polygeist/tools/cgeist/Lib/CGExpr.cc (hunk 3 above)
Comment (etiotto) [reply polygeist/tools/cgeist/Lib/CGExpr.cc:1556]: I added the DEBUG macro... I guess I *should* have used LLVM_DEBUG... oh well.

**Review by etiotto — COMMENTED** [review]
Suggest using LLVM_DEBUG macro to wrap debug code (even is an assertion will be emitted).

**Review by etiotto — APPROVED** [review]

**Pull request merged**